main.c

```diff
@@ -486,22 +430,76 @@ int main(void) {
   halInit();
   chSysInit();
 
-  profiles[currentProfile].callback(ledColors, ledIntensity);
+  /* profiles[currentProfile].callback(ledColors, ledIntensity); */
   updateAnimationSpeed();
 
   // Setup masks to all be 0xFF at the start
   memset(ledMasks, 0xFF, sizeof(ledMasks));
 
+  chBSemObjectInit(&ledDisabledSem, true);
+
   palClearLine(LINE_LED_PWR);
   sdStart(&SD1, &usart1Config);
 
-  // Setup Column Multiplex Timer
-  gptStart(&GPTD_BFTM0, &bftm0Config);
-  gptStartContinuous(&GPTD_BFTM0, SERIAL_CHECK_FREQUENCY);
+  chThdCreateStatic(waThread1, sizeof(waThread1), HIGHPRIO, MsgHandlerThd,
+                    NULL);
 
-  // idle loop, copied form lib/chibios/os/rt/src/chsys.c
   while (true) {
-    port_wait_for_interrupt();
-    CH_CFG_IDLE_LOOP_HOOK();
+
+    // If leds are disabled or enabled and using a reactive profile which does
+    // not need any update, suspend the thread on the semaphore. The semaphore
+    // is flagged up in MsgHandlerThd when leds are enabled or user clicks any
+    // button with the reactive profile enabled.
+    if (!ledEnabled && !reactiveNeedsUpdate) {
+      chBSemWait(&ledDisabledSem);
```

Comment:
This suspends the current thread if the semaphore is owned. `MsgHandlerThd` thread is also suspended waiting for new messages. So it should improve the battery usage as we are not constantly looping.